### ISSUE TITLE
Copying many files is very slow

### ISSUE THREAD

**Issue by dbischof90** [sxyazi/yazi#507]
### What system are you running Yazi on?

Linux Wayland

### What terminal are you running Yazi in?

Alacritty

### Yazi version

yazi 0.1.5

### Did you try the latest main branch to see if the problem has already been fixed?

Tried, but the problem is still present

### Describe the bug

It's not really a bug but maybe a discussion.

For me, yazi does not seem to be fast but responsive - this scales very well but I have the feeling that the actual disk-I/O performance is degrading when copying a lot of files at the same time.

I am coyping files from an external HDD to my local NVME disk via USB3. The remote file system is `ext4`, the local file system is `btrfs`. I'm currently copying about 450.000 files from the HDD, about 300GB of files. I have started the process about 3h ago and am now at about 30% - I have the feeling that if I had copy/pasted everything with `mv/cp` I would have been done by now twice already. 

That's not a bug but maybe some consideration - it looks like in this copy job about 450.000 tasks are created and processed concurrently. I know what you're using `tokio::fs` under the hood and that currently `tokio::fs` still just calls `std::fs` jobs in a separated threadpool and since these tasks are somewhat light but not really "super light" this might build up to quite some performance degradation. The CPU usage for the `yazi`thread hovers around 15% on one core (somewhat recent Ryzen Pro 7), hence my suspicion that the CPU is maybe doing a lot more than it needs to - by jumping around hundreds of thousands of tasks that all call back into a threadpool and thus creating a lot of buzz that slows everything down a bit).

So by all means `yazi` seems like a great application but is it possible that for these "large cases" there could be some optimization to limit the context switching, e.g. some chunking? Breaking up the job into 1-5k and thus limiting the amount of context switching here a bit? Or are there things like ìo_uring` that could improve this?

### Expected Behavior

Performance degradation to unix tools by around 50% (which would be very good already)

### To Reproduce

Slap a huge amount of files to a disk and copy them down

### Configuration

All standard.

### Anything else?

I mentioned `btrfs` - I don't want to exclude this to play a role, my experience with this FS is not that large yet but just eyeballing phoronix benchmarks this should not be such a killer. I saw a lot of files making progress at the same time and the CPU being surprisingly busy, hence my thinking that it could be that.

**Comment by sxyazi** [comment]
Disclaimer: I have no experience with btrfs.

Thank you for the report; I think it makes sense. Yazi has to do more work than `cp`, as it requires semi-real-time reading and summarizing the progress of each task to provide progress reports. This usually means more CPU and IO.

However, please try the latest main branch of Yazi to see if there are any improvements - I've [rewritten the file system](https://github.com/sxyazi/yazi/pull/401) and [refactored the task scheduling system](https://github.com/sxyazi/yazi/pull/462) in the last two months. Then, increase (or decrease) the number of concurrent tasks in your `yazi.toml` config file:

https://github.com/sxyazi/yazi/blob/560a1bf3f9c056f79bbf10b87095e9fe7294fdca/yazi-config/preset/yazi.toml#L70-L72

The specific value depends on your situation, and if it's eventually reduced to 1, it would be similar to `cp`, with only one copy task going on at a time. Maybe we can start by testing it with a small amount of data compared to cp's specific performance gap, such as 5-10G.

Also I want to know if it's related to the copy-on-write mechanism of btrfs? AFAIK, `cp` will try to use copy-on-write, but Rust does not support it (can be implemented through https://crates.io/crates/reflink-copy), which can significantly impact speed.

For such large tasks, I believe directly invoking `cp` is the fastest way - it doesn't need to summarize task progress while copying. The upcoming Yazi v0.2.1 will allow users to delegate Yazi's copying and moving to any external programs they like through Lua plugins.

**Comment by sxyazi** [comment]
Update:

Here are the test results of copying a directory of 13.46 GB to a FAT32 external hard drive using `yazi` and `cp`. I used the following configuration:

- Latest main branch
- ```toml
  [tasks]
  micro_workers    = 10
  macro_workers    = 30
  bizarre_retry    = 5
  ```

## Directory size:

13,287,901,023 bytes (13.46 GB on disk) for 45,402 items

## Using `cp`:

- The entire operation took `4:03`
- CPU usage was around `~40%`, on my MacBook Air M2

![1](https://github.com/sxyazi/yazi/assets/17523360/2dc45ce9-6ecc-40c9-8db8-58b8aff962e9)


## Using `yazi`:

- The entire operation took `2:49`
- CPU usage was around `~70%`, on my MacBook Air M2

![2](https://github.com/sxyazi/yazi/assets/17523360/3acd6868-7787-4e63-bb0e-ee05aff9e5d4)

## Conclusion

Indeed, `yazi` is faster than `cp`, but it consumes more CPU. But that's not all!

I profiled the CPU usage for `yazi` and found that almost all CPU resources are consumed by UI rendering rather than the copying or task scheduling itself:

![3](https://github.com/sxyazi/yazi/assets/17523360/38cc9f63-4293-4157-b795-5e9eae242bf8)

This is because during the copying, frequent UI rendering is required to update the progress bar display in the bottom right corner.

In other words, there is significant optimization potential - we can limit the rendering of the progress bar component to the terminal without triggering a complete UI redraw and it can be optimized in future pull requests.

**Comment by sxyazi** [comment]
Hey, I made a PR to fix it, https://github.com/sxyazi/yazi/pull/509, let me know if it works for you!
